### PR TITLE
Added possibility to alias commands.

### LINKED ISSUE #4172: Add alias support for plugin commands to the PluginManager
# This is a Feature Proposal

## Description

The plugin manager should support `alias` so that the CLI usage can be expressed through different commands (e.g. `serverless deploy function` or `serverless function deploy`).

@HyperBrain has already done some great research on this. The implementation proposal (and some more discussion around that) can be found here: https://github.com/serverless/serverless/issues/4135#issuecomment-323974007

Related to https://github.com/serverless/serverless/issues/4135

### PR DESCRIPTION
**Pull request by HyperBrain**
## What did you implement:

Closes #4172 

## How did you implement it:

* **Command alias support**
Commands now support an alias array that let's them define an alias through which the command can be invoked alternatively. The pluginManager will build an command alias map, that contains all defined aliases with their specific targets.
Aliases are simply defined by adding an `aliases` array to any command, that specifies the command path that will invoke the command as alias, e.g. `install:plugin` would expose the command containing the alias as `serverless install plugin`.
Aliases appear with `serverless help` with the same description as their original commands.

* **Error detection**
The pluginManager will now detect if aliases are specified somewhere and checks for the following restrictions:
  * Aliases cannot overwrite commands and vice versa.
  * Aliases cannot overwrite other aliases

* Added debug logging on command/plugin load time (when SLS_DEBUG is set)

## How can we verify it:

Add an alias to any command like this:
```
deploy:
  ...
  commands: {
    function: {
      usage: 'Deploy a single function from the service',
      lifecycleEvents: [
        'initialize',
        'packageFunction',
        'deploy',
      ],
      aliases: [
        'function:deploy',
      ],
      options: {
        function: {
          usage: 'Name of the function',
          shortcut: 'f',
          required: true,
        },
  ...
```
This sample will expose the `deploy function` command additionally as `function deploy`.

`serverless help` should show `function deploy` now as available command.

Invocation of `serverless` should fail, if you set the alias to an existing command, so you're aware of the error immediately when implementing commands or changing their configuration.
BTW: The unit tests check for all possible error cases.

## Todos:

- [x] Write tests
- [ ] Write documentation
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config / commands / resources
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO

### CODE REVIEW

File: lib/classes/PluginManager.test.js [hunk 1]
```diff
@@ -377,7 +379,7 @@ describe('PluginManager', () => {
     let getCommandsStub;
 
     beforeEach(() => {
-      writeFileStub = sinon.stub().resolves();
+      writeFileStub = sinon.stub().returns(BbPromise.resolve());
```
Comment (HyperBrain) [inline lib/classes/PluginManager.test.js]: @pmuens This is critical. Someone added a `stub().resolves()` here, but this is not available in sinon 1.x which is used in Serverless. I fixed it by using `.returns(...)`. We should immediately update to sinon 2.x because people are already used to it, and as long as the unit tests are completely broken, noone will ever see the issues and the codebase deteriorates and gets bad.

**Comment by HyperBrain** [comment]
@pmuens @horike37 The command aliases are fully functional and unit tested now. 
However, I did not add them to the plugin author's guide, because I think that needs a rewrite anyway, as it currently is only a very high level documentation that even lacks entrypoint documentation.

File: lib/classes/PluginManager.test.js (hunk 1 above)
Comment (pmuens) [reply lib/classes/PluginManager.test.js]: @HyperBrain Really, really nice catch 👍

I used `resolves()` for the last couple of months w/o noticing that we're running such an old `sinon` version! Yes, we should definitely update it ASAP!

**Comment by pmuens** [comment]
> @pmuens @horike37 The command aliases are fully functional and unit tested now.
However, I did not add them to the plugin author's guide, because I think that needs a rewrite anyway, as it currently is only a very high level documentation that even lacks entrypoint documentation.

Great! Thanks for implementing this @HyperBrain 👍 💯. It's a really nifty feature. Yes, we could add the documentation later on once it's "polished"...

Weil take a look at it ASAP.

File: lib/classes/PluginManager.test.js (hunk 1 above)
Comment (horike37) [reply lib/classes/PluginManager.test.js]: sinon v3.x has been already released.
http://sinonjs.org/releases/index.html

We should update to follow the migration guide. 
Here are the steps.
http://sinonjs.org/releases/v3.2.1/migrating-to-2.0/
http://sinonjs.org/releases/v3.2.1/migrating-to-3.0/

Comment (pmuens) [reply lib/classes/PluginManager.test.js]: Thanks for the links @horike37 👍

I just figured that this `resolves` should work since we use `sinon-bluebird`. However we should update `sinon` ASAP (we're currently on v1 and v3 is the latest one).

I started the update in https://github.com/serverless/serverless/pull/4200 (but it's currently on hold since we need a way to use Bluebird rather than the native Promise implementation).

Comment (HyperBrain) [reply lib/classes/PluginManager.test.js]: @pmuens `sinon-bluebird` is deprecated with version 2.

Regarding the promise type used by sinon. It's quite easy to set the implementation:
`sinon.usingPromise(BbPromise.Promise)`

**Comment by HyperBrain** [comment]
@pmuens Did you add the `deploy` alias in a plugin before the deploy plugin? The unit tests check this case, so it would be great if you post the specific command declaration. Then I can check what exactly happens.
Specifically the code at pluginManager@133 should throw an error in that case:
```
    if (_.reduce(splitAlias, (__, aliasPath) => {
      if (!__ || !__.commands || !__.commands[aliasPath]) {
        return null;
      }
      return __.commands[aliasPath];
    }, this)) {
      throw new this.serverless.classes
        .Error(`Command ${alias} cannot be overriden by an alias`);
    }
```
In case you define an alias first, and then a command is applied, the command creation will fail in a similar way (as there is already an alias). So either way Serverless should report an error.

**Comment by HyperBrain** [comment]
Oh wait, I think I found the issue wight now ;-) Fixing

**Comment by HyperBrain** [comment]
@pmuens Hmmm... I added `aliases: [ 'deploy' ]` to the info plugin for testing. A `serverless help` bails out with an error then:
```
$ SLS_DEBUG=* node node_modules/serverless/bin/serverless help
Serverless: Load command run
Serverless: Load command config
Serverless: Load command config:credentials
Serverless: Load command create
Serverless: Load command install
Serverless: Load command package
Serverless: Load command deploy
Serverless: Load command deploy:function
Serverless: Load command deploy:list
Serverless: Load command deploy:list:functions
Serverless: Load command invoke
Serverless: Load command invoke:local
Serverless: Load command info
Serverless:   -> @deploy

  Serverless Error ---------------------------------------

  Command deploy cannot be overriden by an alias
```

Which is exactly the expected error message. I need the exact setup. Maybe you triggered some edge case here.

**Comment by pmuens** [comment]
Thanks for getting back @HyperBrain 👍

Sorry, that I haven't included steps to reproduce the problem. I just tested it again today. Here are my findings:

This works and shows the error you've described in https://github.com/serverless/serverless/pull/4198#issuecomment-327133494:

```javascript
class Info {
  constructor(serverless) {
    this.serverless = serverless;

    this.commands = {
      info: {
        usage: 'Display information about the service',
        lifecycleEvents: [
          'info',
        ],
        aliases: [
          'deploy',
        ],
        options: {
          stage: {
            usage: 'Stage of the service',
            shortcut: 's',
          },
          region: {
            usage: 'Region of the service',
            shortcut: 'r',
          },
          verbose: {
            usage: 'Display Stack output',
            shortcut: 'v',
          },
        },
      },
    };

    // ...snip...
  }

  // ...snip...
}
```

This works fine and doesn't print an error (so it's the edge case when you add an alias to the same plugin but in a sub command):

```javascript
class Deploy {
  constructor(serverless, options) {
    this.serverless = serverless;
    this.options = options || {};

    Object.assign(
      this,
      validate
    );

    this.commands = {
      deploy: {
        usage: 'Deploy a Serverless service',
        lifecycleEvents: [
          'deprecated#cleanup->package:cleanup',
          'deprecated#initialize->package:initialize',
          'deprecated#setupProviderConfiguration->package:setupProviderConfiguration',
          'deprecated#createDeploymentArtifacts->package:createDeploymentArtifacts',
          'deprecated#compileFunctions->package:compileFunctions',
          'deprecated#compileEvents->package:compileEvents',
          'deploy',
          'finalize',
        ],
        options: {
          stage: {
            usage: 'Stage of the service',
            shortcut: 's',
          },
          region: {
            usage: 'Region of the service',
            shortcut: 'r',
          },
          package: {
            usage: 'Path of the deployment package',
            shortcut: 'p',
          },
          verbose: {
            usage: 'Show all stack events during deployment',
            shortcut: 'v',
          },
          force: {
            usage: 'Forces a deployment to take place',
          },
        },
        commands: {
          function: {
            usage: 'Deploy a single function from the service',
            lifecycleEvents: [
              'initialize',
              'packageFunction',
              'deploy',
            ],
            // HERE'S THE ALIAS DEFINITION
            aliases: [
              'deploy',
            ],
            options: {
              function: {
                usage: 'Name of the function',
                shortcut: 'f',
                required: true,
              },
              stage: {
                usage: 'Stage of the function',
                shortcut: 's',
              },
              region: {
                usage: 'Region of the function',
                shortcut: 'r',
              },
              force: {
                usage: 'Forces a deployment to take place',
              },
            },
          },
          list: {
            usage: 'List deployed version of your Serverless Service',
            lifecycleEvents: [
              'log',
            ],
            commands: {
              functions: {
                usage: 'List all the deployed functions and their versions',
                lifecycleEvents: [
                  'log',
                ],
              },
            },
          },
        },
      },
    };

    // ...snip...
}
```

Running `serverless help` prints the following:

bildschirmfoto 2017-09-06 um 13 15 18

**Comment by HyperBrain** [comment]
@pmuens Thanks for the explanation 👍 . I will add handling for the edge case, so that it will detect invalid overrides in the current command too.

Just saw that I have to rebase and adjust the unit tests due to conflicts. Will do that too.

**Comment by pmuens** [comment]
> @pmuens Thanks for the explanation 👍 . I will add handling for the edge case, so that it will detect invalid overrides in the current command too.

> Just saw that I have to rebase and adjust the unit tests due to conflicts. Will do that too.

Great! Thanks for updating this @HyperBrain 👍. Let us know if you need any help with that.

**Comment by HyperBrain** [comment]
@pmuens The functionality is fixed now and prohibits aliasing of commands in the same plugin now. Additionally I fixed a small bug that has hidden commands and aliases in the help screen. 

See this example where I added 2 aliases to `function deploy`:
image

Additionally your example now throws the correct error:
```
Serverless: Load command run
Serverless: Load command config
Serverless: Load command config:credentials
Serverless: Load command create
Serverless: Load command install
Serverless: Load command package
Serverless: Load command deploy
Serverless: Load command deploy:function
Serverless:   -> @deploy

  Serverless Error ---------------------------------------

  Command "deploy" cannot be overriden by an alias
```

**Comment by HyperBrain** [comment]
@pmuens Rebased and resolved the conflicts. Again ready for merge from my side 🍄

**Comment by pmuens** [comment]
> @pmuens Rebased and resolved the conflicts. Again ready for merge from my side 🍄

Nice! Thanks for fixing that @HyperBrain 👍. I'll take another deep dive into this tomorrow morning!

**Comment by HyperBrain** [comment]
Just added some unit tests to make Coveralls happy again 😄

**Comment by pmuens** [comment]
> Just added some unit tests to make Coveralls happy again 😄

😄 looks like this coveralls thing is working as expected 😂...

**Comment by eahefnawy** [comment]
I've followed @pmuens edge cases and couldn't break it on my machine either. Awesome stuff @HyperBrain 😊 

Just one note, I know we don't have comprehensive plugin development docs, but I feel that after merging this in, no plugin author would know about this feature. I think most of our plugin authors follow from our own internal plugins, and none of our internal plugins use this alias commands yet.

I think it's best if we add some notes about this in the plugin guide in our docs?

What you guys think?

**Comment by HyperBrain** [comment]
@eahefnawy I had a look at the current plugin author's docs recently and I'm afraid, that this feature is not the only one the readers will miss. In my opinion the guide needs a complete rewrite 😄 .

**Comment by eahefnawy** [comment]
@HyperBrain lol agreed!

**Comment by eahefnawy** [comment]
@HyperBrain You've done tons of awesome enhancements to our plugin system that is not covered in the docs. Interested in writing a blog post on our official blog about those new enhancements (including this one)?

Totally understood if writing is not your thing of course ☺️

**Comment by HyperBrain** [comment]
@eahefnawy I can do that and wrap-up all the recent changes in the plugin system. This would be about deprecation, aliases, spawning as delegation (like the `deploy --function`), and I might find other things that changed.

**Comment by eahefnawy** [comment]
Awesome! Let's sync on Slack!

Merging... 💥
